File: src/types/parcel.ts
import type { Deal } from './deal';
import type { GraphEntity, GraphRelatedEntity } from './graph';

export type ParcelReviewStatus = 'candidate' | 'shortlisted' | 'dismissed';
export type AcquisitionCaseStatus = ParcelReviewStatus | 'contacted' | 'promoted';
export type AcquisitionActivityType = 'call' | 'email' | 'sms' | 'meeting' | 'note';
export type ParcelPersona = 'developer' | 'investor' | 'broker' | 'realtor';

export interface ParcelFact {
  id: string;
  fact_type: string;
  value: unknown;
  source_url?: string | null;
  field_path?: string | null;
  excerpt?: string | null;
  confidence: number;
  observed_at: string;
  last_verified_at: string;
}

export type ParcelLineageEventType = 'split' | 'merge' | 'replat' | 'correction';

export interface ParcelLineageEvidence {
  id: string;
  raw_source_record_id: string;
  source_url?: string | null;
  excerpt?: string | null;
  confidence: number;
  observed_at: string;
  last_verified_at: string;
  payload?: Record<string, unknown> | null;
}

export interface ParcelLineageParticipant {
  id: string;
  role: 'predecessor' | 'successor';
  external_parcel_id: string;
  parcel_id?: string | null;
  parcel?: ParcelSummary | null;
  last_verified_at: string;
}

export interface ParcelLineageEvent {
  id: string;
  source_key: string;
  external_event_id: string;
  event_type: ParcelLineageEventType;
  confidence: number;
  observed_at: string;
  last_verified_at: string;
  attributes?: Record<string, unknown> | null;
  participants: ParcelLineageParticipant[];
  evidence: ParcelLineageEvidence[];
}

export interface ParcelSummary {
  id: string;
  external_parcel_id: string;
  parcel_group_id?: string | null;
  jurisdiction?: string | null;
  county?: string | null;
  state?: string | null;
  address?: string | null;
  city?: string | null;
  postal_code?: string | null;
  latitude: number;
  longitude: number;
  land_area_sq_ft?: number | null;
  improvement_area_sq_ft?: number | null;
  land_value?: number | null;
  improvement_value?: number | null;
  total_assessed_value?: number | null;
  land_use?: string | null;
  zoning_code?: string | null;
  boundary_geometry?: Record<string, unknown> | null;
  last_verified_at: string;
}

export interface NearbyParcelCandidate {
  id: string;
  rank: number;
  distance_miles: number;
  score: number;
  score_confidence: number;
  explanation: {
    reasons?: string[];
    cautions?: string[];
    [key: string]: unknown;
  };
  review_status: ParcelReviewStatus;
  assigned_to_user_id?: string | null;
  assigned_to_name?: string | null;
  assigned_by_user_id?: string | null;
  assigned_at?: string | null;
  ranker_version: string;
  parcel: ParcelSummary;
  facts: ParcelFact[];
}

export interface NearbyParcelSearchSummary {
  id: string;
  deal_id: string;
  anchor_brand_match_id?: string | null;
  anchor_permit_id: string;
  anchor_latitude: number;
  anchor_longitude: number;
  radius_miles: number;
  persona: ParcelPersona;
  filters?: Record<string, unknown> | null;
  result_limit: number;
  as_of: string;
  ranker_version: string;
  created_at: string;
}

export interface NearbyParcelSearch extends NearbyParcelSearchSummary {
  candidates: NearbyParcelCandidate[];
}

export interface NearbyParcelSearchCreate {
  anchor_brand_match_id: string;
  radius_miles: number;
  persona: ParcelPersona;
  limit?: number;
  minimum_land_area_sq_ft?: number | null;
  zoning_codes?: string[];
  land_uses?: string[];
}

export interface NearbyParcelCandidateAssignment {
  assigned_to_user_id: string;
}

export interface NearbyParcelOpportunityCreate {
  name?: string | null;
}

export interface NearbyParcelOpportunityResponse {
  created: boolean;
  candidate_id: string;
  deal: Deal;
}

export interface ParcelSearchHit {
  search_id: string;
  deal_id: string;
  deal_name?: string | null;
  persona: ParcelPersona;
  radius_miles: number;
  created_at: string;
  rank: number;
  distance_miles: number;
  score: number;
  score_confidence: number;
  review_status: ParcelReviewStatus;
}

export interface ParcelDetail {
  parcel: ParcelSummary;
  facts: ParcelFact[];
  search_count: number;
  search_hits: ParcelSearchHit[];
  graph_entity?: GraphEntity | null;
  graph_related: GraphRelatedEntity[];
  lineage_events: ParcelLineageEvent[];
}

export interface AcquisitionRadarSignal {
  candidate_id: string;
  search_id: string;
  deal_id: string;
  deal_name: string;
  persona: ParcelPersona;
  approval_stage?: string | null;
  signal_confidence?: number | null;
  distance_miles: number;
  candidate_score: number;
  created_at: string;
}

export interface AcquisitionRadarItem {
  parcel: ParcelSummary;
  candidate_id: string;
  acquisition_case_id?: string | null;
  radar_score: number;
  best_candidate_score: number;
  score_confidence: number;
  appearance_count: number;
  opportunity_count: number;
  personas: ParcelPersona[];
  review_status: AcquisitionCaseStatus;
  assigned_to_user_id?: string | null;
  assigned_to_name?: string | null;
  contacted_at?: string | null;
  follow_up_at?: string | null;
  promoted_deal_id?: string | null;
  latest_signal_at: string;
  reasons: string[];
  cautions: string[];
  signals: AcquisitionRadarSignal[];
}

export interface AcquisitionRadarResponse {
  items: AcquisitionRadarItem[];
  total: number;
  limit: number;
  offset: number;
  summary: {
    total_parcels: number;
    shortlisted_parcels: number;
    multi_opportunity_parcels: number;
    assigned_parcels: number;
    state_count: number;
  };
}

export interface AcquisitionRadarParams {
  q?: string;
  state?: string;
  persona?: ParcelPersona;
  review_status?: AcquisitionCaseStatus;
  assignment?: 'assigned' | 'unassigned';
  limit?: number;
  offset?: number;
}

export interface ParcelAcquisitionSource {
  id: string;
  candidate_id: string;
  search_id: string;
  created_at: string;
}

export interface ParcelAcquisitionActivity {
  id: string;
  activity_type: AcquisitionActivityType;
  notes?: string | null;
  actor_user_id?: string | null;
  occurred_at: string;
  follow_up_at?: string | null;
  created_at: string;
}

export interface ParcelAcquisitionCase {
  id: string;
  parcel_id: string;
  status: AcquisitionCaseStatus;
  assigned_to_user_id?: string | null;
  assigned_to_name?: string | null;
  assigned_by_user_id?: string | null;
  assigned_at?: string | null;
  contacted_at?: string | null;
  follow_up_at?: string | null;
  promoted_deal_id?: string | null;
  created_at: string;
  updated_at: string;
  sources: ParcelAcquisitionSource[];
  activities: ParcelAcquisitionActivity[];
}

export interface ParcelAcquisitionCaseUpdate {
  status?: AcquisitionCaseStatus;
  assigned_to_user_id?: string | null;
  follow_up_at?: string;
}

export interface ParcelAcquisitionActivityCreate {
  activity_type: AcquisitionActivityType;
  notes?: string;
  occurred_at?: string;
  follow_up_at?: string;
}

File: app/schemas/parcel.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.deal import DealDetailResponse


class NearbyParcelSearchCreate(BaseModel):
    anchor_brand_match_id: str
    radius_miles: float = Field(default=2.0, ge=0.25, le=5.0)
    persona: Literal["developer", "investor", "broker", "realtor"] = "developer"
    minimum_land_area_sq_ft: Optional[float] = Field(default=None, ge=0)
    zoning_codes: list[str] = Field(default_factory=list, max_length=50)
    land_uses: list[str] = Field(default_factory=list, max_length=50)
    limit: int = Field(default=50, ge=1, le=100)


class NearbyParcelCandidateReview(BaseModel):
    review_status: Literal["candidate", "shortlisted", "dismissed"]


class NearbyParcelCandidateAssignment(BaseModel):
    assigned_to_user_id: str = Field(..., min_length=1)


class NearbyParcelOpportunityCreate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)


class ParcelFactResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    fact_type: str
    value: Any
    source_url: Optional[str]
    field_path: Optional[str]
    excerpt: Optional[str]
    confidence: float
    observed_at: datetime
    last_verified_at: datetime


class ParcelSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    external_parcel_id: str
    parcel_group_id: Optional[str]
    jurisdiction: Optional[str]
    county: Optional[str]
    state: Optional[str]
    address: Optional[str]
    city: Optional[str]
    postal_code: Optional[str]
    latitude: float
    longitude: float
    land_area_sq_ft: Optional[float]
    improvement_area_sq_ft: Optional[float]
    land_value: Optional[float]
    improvement_value: Optional[float]
    total_assessed_value: Optional[float]
    land_use: Optional[str]
    zoning_code: Optional[str]
    boundary_geometry: Optional[dict[str, Any]] = None
    last_verified_at: datetime


class NearbyParcelCandidateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    rank: int
    distance_miles: float
    score: float
    score_confidence: float
    explanation: dict[str, Any]
    review_status: str
    assigned_to_user_id: Optional[str] = None
    assigned_to_name: Optional[str] = None
    assigned_by_user_id: Optional[str] = None
    assigned_at: Optional[datetime] = None
    ranker_version: str
    parcel: ParcelSummaryResponse
    facts: list[ParcelFactResponse] = Field(default_factory=list)


class NearbyParcelSearchSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    deal_id: str
    anchor_brand_match_id: Optional[str]
    anchor_permit_id: str
    anchor_latitude: float
    anchor_longitude: float
    radius_miles: float
    persona: str
    filters: Optional[dict[str, Any]]
    result_limit: int
    as_of: datetime
    ranker_version: str
    created_at: datetime


class NearbyParcelSearchResponse(NearbyParcelSearchSummary):
    candidates: list[NearbyParcelCandidateResponse] = Field(default_factory=list)


class ParcelSearchHitResponse(BaseModel):
    search_id: str
    deal_id: str
    deal_name: Optional[str] = None
    persona: str
    radius_miles: float
    created_at: datetime
    rank: int
    distance_miles: float
    score: float
    score_confidence: float
    review_status: str


class ParcelDetailResponse(BaseModel):
    parcel: ParcelSummaryResponse
    facts: list[ParcelFactResponse] = Field(default_factory=list)
    search_count: int = 0
    search_hits: list[ParcelSearchHitResponse] = Field(default_factory=list)
    graph_entity: Optional[dict[str, Any]] = None
    graph_related: list[dict[str, Any]] = Field(default_factory=list)


class NearbyParcelOpportunityResponse(BaseModel):
    created: bool
    candidate_id: str
    deal: DealDetailResponse


class AcquisitionRadarSignalResponse(BaseModel):
    candidate_id: str
    search_id: str
    deal_id: str
    deal_name: str
    persona: str
    approval_stage: Optional[str] = None
    signal_confidence: Optional[float] = None
    distance_miles: float
    candidate_score: float
    created_at: datetime


class AcquisitionRadarItemResponse(BaseModel):
    parcel: ParcelSummaryResponse
    candidate_id: str
    radar_score: float
    best_candidate_score: float
    score_confidence: float
    appearance_count: int
    opportunity_count: int
    personas: list[str]
    review_status: str
    assigned_to_user_id: Optional[str] = None
    assigned_to_name: Optional[str] = None
    latest_signal_at: datetime
    reasons: list[str]
    cautions: list[str]
    signals: list[AcquisitionRadarSignalResponse]


class AcquisitionRadarSummaryResponse(BaseModel):
    total_parcels: int
    shortlisted_parcels: int
    multi_opportunity_parcels: int
    assigned_parcels: int
    state_count: int


class AcquisitionRadarResponse(BaseModel):
    items: list[AcquisitionRadarItemResponse]
    total: int
    limit: int
    offset: int
    summary: AcquisitionRadarSummaryResponse
